store errors in the job run and show to the user

peetucket opened the pull request:
## Why was this change made? 🤔

Todo:
- [x] run integration tests
- [ ] validate with PO

Fixes #945 and fixes #936:

 - store any errors that occurred during processing in the database so we can show to the user (for both pre-assembly and discovery jobs)
 - put discovery report and pre-assembly jobs into an error or failure state based on if individual objects had an error or the whole job failed


Example error message display to user on detail page:

![Screen Shot 2022-06-08 at 11 54 14 AM](https://user-images.githubusercontent.com/47137/172694750-3494c552-db94-4e82-a45c-6ac2747231bc.png)


## How was this change tested? 🤨

New tests

peetucket commented on app/jobs/preassembly_job.rb:
```diff
@@ -6,8 +6,13 @@ class PreassemblyJob < ApplicationJob
+    else
+      job_run.completed
+    end
   end
```
this changes around how we report back errors to the job ... instead of returning a boolean from `batch#run_pre_assembly` to indicate errors/no errors, use a new accessor on the `batch` object to indicate if we had errors, and then fetch the error message from the object to store in the job_run if so

peetucket commented on app/lib/pre_assembly/batch.rb:
```diff
@@ -121,14 +123,17 @@ def process_digital_objects
+      errors.each { |error| log "**WARNING**: #{error}" }
+      @had_errors = !errors.size.zero? # indicate if we had any errors
+      @error_message = errors.join(', ') # set the error message so they can be saved in the job_run
 
```
we will now store the error messages and if any errors occurred in the accessor so the caller of this method can introspect and see what happened

peetucket commented on app/views/job_runs/show.html.erb:
```diff
@@ -24,6 +24,10 @@
+      <dt class="col-sm-2">Errors</dt>
+      <dd class="col-sm-10"><%= @job_run.error_message%></dd>
+    <% end %>
     <dt class="col-sm-2">Job Output Log</dt>
```
show error messages to the user on the detail page if they exist

peetucket commented on spec/lib/pre_assembly/batch_spec.rb:
```diff
@@ -28,9 +28,9 @@
       expect(batch).to receive(:process_digital_objects)
-      expect(batch.run_pre_assembly).to be true
+      batch.run_pre_assembly
     end
```
no longer returning a true/false from the method

peetucket commented on app/jobs/discovery_report_job.rb:
```diff
@@ -6,10 +6,15 @@ class DiscoveryReportJob < ApplicationJob
+      job_run.error_message = e.message
+      job_run.failed
+    end
   end
```
We will now rescue any overall job exceptions and put the job into a failed state with the error message we can show to the user.  We can use this later to raise exceptions for things like bad manifests and have them caught here and reported to the user.

peetucket commented on app/lib/pre_assembly/batch.rb:
```diff
@@ -98,10 +99,11 @@ def load_checksums(dobj)
     # rubocop:disable Metrics/MethodLength
-    # @return [boolean] true if all objects succeeded, false if any errors
+    # rubocop:disable Metrics/CyclomaticComplexity
     def process_digital_objects
```
I created ticket #957 to simplify this method so we can remove some/all of these rubocop exceptions, but didn't want to do too much in this PR

peetucket commented on app/services/discovery_report.rb:
```diff
@@ -6,6 +6,7 @@
 class DiscoveryReport
   attr_reader :batch, :start_time, :summary
+  attr_accessor :had_errors, :error_message
 
```
same theory as above, we will track errors and messages for discovery reports too

peetucket commented on app/services/discovery_report.rb:
```diff
@@ -29,18 +30,24 @@ def each_row
       # log the output to a running progress file
-      File.open(batch.batch_context.progress_log_file, 'a') { |f| f.puts log_progress_info(dobj).to_yaml }
+      File.open(batch.batch_context.progress_log_file, 'a') { |f| f.puts log_progress_info(dobj, status).to_yaml }
       yield row
```
log the state of each object

peetucket commented on app/lib/pre_assembly/batch.rb:
```diff
@@ -121,14 +123,17 @@ def process_digital_objects
 
-      num_failures == 0 # if no failures in any objects, return true, else false
+      log "#{total_obj} objects pre-assembled"
     end
```
we will now store the error messages and if any errors occurred in the accessor so the caller of this method can introspect and see what happened

peetucket commented on app/services/discovery_report.rb:
```diff
@@ -70,9 +71,12 @@ def file_manifest
+    @had_errors = (@summary[:objects_with_error] > 0) # indicate if any objects generated errors
+    @error_message = "#{@summary[:objects_with_error]} objects had errors in the discovery report" if @had_errors
+    json_report
   end
```
if the report had any errors for any objects, record that and the error message, so we can store that in the job run table in the database and set the job run into the correct state

peetucket commented on spec/services/discovery_report_spec.rb:
```diff
@@ -29,12 +38,6 @@
-      # In the actual application, `batch_context.output_dir_no_exists!` would get run and thus we would always have a unique folder created for each job
-      FileUtils.mkdir_p(batch.batch_context.output_dir)
-      FileUtils.rm_f(batch.batch_context.progress_log_file)
       expect(report).to receive(:process_dobj).with(dig_obj1).and_return(validator1)
```
this is useful for our new test below, so move it to a before block

ndushay commented on app/services/discovery_report.rb:
```diff
@@ -76,9 +77,12 @@ def file_manifest
       json.rows { json.array!(each_row) }
       json.summary summary
     end
+    @objects_had_errors = (@summary[:objects_with_error] > 0) # indicate if any objects generated errors
```
could this be `summary` instead of `@summary`?   I lose track of when we need the `@` ... but maybe just when assigning?

ndushay reviewed (approved):
This looks good, and all of my comments are quibbles and perhaps shouldn't hold up this PR (could be a followon):

- do any of our tests run multiple objects through so you have the case of erroring on some objects and succeeding on some objects in a single batch?   (I noticed for the feature tests, all of the batches in use were single object)  This seems useful in a test for each job type, no?
- there is a separate ticket to write feature tests, and i would think we would want, for both flavors of jobs, feature tests for "failure" and "completed with errors"  ~~I will update that ticket.~~ #942, updated.  (#942 is marked as blocked until this PR is merged.)
- I'm wondering if `@` is needed as often as you are using it ... I thought the accessors meant we didn't need `@` for gets ... but :shrug: ?   Obviously not crucial.
- a few comment quibbles, can be ignored.

ndushay commented:
@peetucket bless you for running integration tests and making any PRs they need (as well as proving that this works in situ)

ndushay commented on app/jobs/discovery_report_job.rb:
```diff
@@ -4,12 +4,23 @@ class DiscoveryReportJob < ApplicationJob
   def perform(job_run)
     job_run.started
     report = job_run.to_discovery_report
+    # the .to_builder is where all the work of the report happens, it iterates over the objects and produces JSON
```
suggest:  `.to_builder produces JSON report by iterating over the objects`

ndushay commented on app/jobs/discovery_report_job.rb:
```diff
@@ -4,12 +4,23 @@ class DiscoveryReportJob < ApplicationJob
+    job_run.output_location = file.path
     job_run.save!
-    job_run.completed
+    if report.objects_had_errors # this is when errors occur on individual objects when running the report
```
suggest: `individual objects processed had errors`

ndushay commented on app/jobs/discovery_report_job.rb:
```diff
@@ -4,12 +4,23 @@ class DiscoveryReportJob < ApplicationJob
+    else
+      job_run.completed
+    end
+  rescue StandardError => e # this catches an exception that occurs on the entire job
```
suggest: `catch any error preventing the whole job from running (e.g. bad header in csv)`

ndushay reviewed (commented):
added comment quibbles;  they apply to both job types, of course.

peetucket commented on app/models/job_run.rb:
```diff
@@ -51,8 +51,13 @@ def mail_subject
   def send_notification
-    return unless output_location
+    return if in_progress?
 
```
guard condition now uses the state machine instead of the database field (which is no longer updated for pre-assembly jobs)

peetucket commented:
Note...since approval: 

1. update code comments with suggestions
2. update use of accessors
3. Fixed an issue with email notifications that weren't going out for pre-assembly jobs (found during integration testsing) and fixed the issue (which had to do with a guard clause in the `JobRun` class that needed updating: https://github.com/sul-dlss/pre-assembly/pull/956/files#diff-477e241400bf18f5f78f9e2154b2b48844e4600d5702b0a2cb7dc8e97c621960R59-R61)

peetucket commented on app/views/shared/_header.html.erb:
```diff
@@ -26,8 +26,8 @@
-      <button type="button" class="close" data-dismiss="alert" aria-label="Close"><i class="fa-solid xmark"></i></button>
+    <div class="alert alert-<%= key %> alert-dismissible fade show" role="alert">
+      <button type="button" class="btn-close" data-bs-dismiss="alert" aria-label="Close"></button>
       <%= value %>
```
Bootstrap 5 update broke the close button on alerts - this fixes it

peetucket commented:
Note that in this PR, spec/services/discovery_report_spec.rb and spec/lib/pre_assembly/batch_spec.rb have mocked data that simulate what happens if one or more objects fail, verifying that `object_had_errors` and `error_message` are set appropriately, and then there are separate tests for the job classes that verify they get into the correct state based on this.  We could consider (perhaps separately?) writing a feature test for the full lifecycle though.

ndushay commented:
@peetucket if this is ready to go, remove [Hold] in the title and feel free to merge.

peetucket commented:
> @peetucket if this is ready to go, remove [Hold] in the title and feel free to merge.

ok will do... just wanted to do a quick review with Andrew first so he understands what is changing, then will merge if no objections from him

ndushay commented:
Can this be merged now?  It would be helpful to both John and me if it were merged as is.  Other work can be in other PRs.

ndushay commented:
So the holdup is:   you can't download the discovery report:

from Andrew on slack:
> The report file is there on the file system and it looks like a complete report. It’s just not downloadable.
> Previously, you would be able to download the report (without the message saying there’s an error), so this would be a regression.

example: https://sul-preassembly-qa.stanford.edu/job_runs/351

peetucket commented:
> So the holdup is: you can't download the discovery report:
> 
> from Andrew on slack:
> 
> > The report file is there on the file system and it looks like a complete report. It’s just not downloadable.
> > Previously, you would be able to download the report (without the message saying there’s an error), so this would be a regression.
> 
> example: https://sul-preassembly-qa.stanford.edu/job_runs/351

This was a small fix -- was only showing the report for jobs that were completed instead of jobs that were any "finished" state (e.g. "completed", "completed (with errors)", "failed").  Fixed this and also ensured that downloads are only offered if the file actually exists (in some cases a "failed" state will result in no actual report).

ndushay commented:
can you deploy to stage or qa where there are discovery report jobs in different states and I'll click around and vet it?

peetucket commented:
> can you deploy to stage or qa where there are discovery report jobs in different states and I'll click around and vet it?

it's on qa and stage now

peetucket commented:
That take back -- use qa for now and hold this for now -- need to deal with stage tomorrow (looks like maybe another database migration is needed to adjust the states of all jobs that are currently done when deploying to a new environment for the first time).

ndushay commented:
As a standin for Andrew, I deployed the branch to qa and looked at reports for completed, completed with errors, and failed and clicked the download links and got the right reports (except I think andrew used the same name twice and the later log overwrote the earlier log).

So I am ✅ on merging this tomorrow.

peetucket commented on app/lib/pre_assembly/batch.rb:
```diff
@@ -8,7 +8,9 @@ class Batch
-                  :file_manifest
+                  :file_manifest,
+                  :error_message,
+                  :objects_had_errors
```
re-ordered
